posts/2019/06/python-library-method-usage.md

```diff
@@ -0,0 +1,217 @@
+<!--
+.. title: Method usage for popular numberical and scientific libraries
+.. slug: python-package-method-usage
+.. date: 2019-06-01
+.. author: Christopher Ostrouchov
+.. tags: Labs
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+-->
+
+Developers of open source software often have a difficult time
+understanding how others use their libraries. Having a better data of
+when and how methods are being used allows for better api design, more
+instructive tutorials, and understanding the adoption of new
+features.
+
+# Python Namespace Inspection
+
+A general tool
+[python-api-inspect](https://github.com/Quansight-Labs/python-api-inspect)
+was written to analyze any method/attribute call within a given set of
+namespaces within a repository. This work was heavily inspired by
+prior work inspecting method usage with [google big
+query](https://galeascience.wordpress.com/2016/08/10/top-10-pandas-numpy-and-scipy-functions-on-github/)
+for pandas, numpy, and scipy. The previously mentioned blog post used
+regular expressions to search for method usage. The primary issues
+with this approach is that is cannot handle `import numpy as nmpy;
+nmpy.array(...)` unless an additional regualr expression is
+constructed for this case. Thus it is not general enough and does not
+scale well with the number of libraries that we would like to inspect
+method and attribute usage for.
+
+A more robust approach is to inspect the python abstract syntax
+tree. Python comes with a performant `import ast; ast.parse(...)`
+method for constructing a python ast from source code. With the ast a
+[node
+visitor](https://docs.python.org/3/library/ast.html#ast.NodeVisitor)
+from the stdlib is used to record `import` statements. This allows us
+to catch any absolute namespace reference. The following are cases
+that
+[python-api-inspect](https://github.com/Quansight-Labs/python-api-inspect)
+catches.
+
+```python
+import numpy
+import numpy as np
+import numpy.random as rnd
+from numpy import random as rand
+
+numpy.array([1, 2, 3])
+numpy.random.random((2, 3))
+np.array([1, 2, 3])
+rnd.random((2, 3))
+rand.random((2, 3))
+```
+
+There are limitations to this approach since python is a heavily duck
+typed language. To understand this see this example.
+
+```python
+a = numpy.array(...)
+a.transpose()
+
+
+def foobar(array):
+    return array.transpose()
+```
+
+How is one supposed to infer that `a.transpose()` is a numpy method or
+`foobar` is a method that takes a numpy array in as input? These are
+open questions that would allow for futher inspection of how libraries
+use given functions and attributes. Now that the internals of the tool
+have been discussed usage is quite simple. The `inspect.py` has heavy
+caching of downloaded repositories and source files that have been
+analyzed. Currently the repository inspects 17 libraries/namespaces
+and around 10,000 repositories.
+
```

Comment:
needs a sentence here like: "The API inspection tool can be run from the command line, ....:"